File: js/src/pricing-tiers.ts
import { Decimal } from "./decimal.js";

export type RateKind =
  | "input"
  | "output"
  | "cache_read"
  | "cache_read_audio"
  | "cache_creation"
  | "cache_creation_audio"
  | "cache_creation_1h"
  | "reasoning"
  | "input_audio"
  | "output_audio"
  | "input_image"
  | "output_image";
export type PricingMode = "standard" | "batch" | "priority" | "flex";

/** Upstream field name for each rate kind. */
export const RATE_FIELDS: Record<RateKind, string> = {
  input: "input_cost_per_token",
  output: "output_cost_per_token",
  cache_read: "cache_read_input_token_cost",
  cache_read_audio: "cache_read_input_audio_token_cost",
  cache_creation: "cache_creation_input_token_cost",
  cache_creation_audio: "cache_creation_input_audio_token_cost",
  cache_creation_1h: "cache_creation_input_token_cost_above_1hr",
  reasoning: "output_cost_per_reasoning_token",
  input_audio: "input_cost_per_audio_token",
  output_audio: "output_cost_per_audio_token",
  input_image: "input_cost_per_image_token",
  output_image: "output_cost_per_image_token"
};

export const MODE_SUFFIXES: Record<PricingMode, string> = {
  standard: "",
  batch: "batches",
  priority: "priority",
  flex: "flex"
};

const RATE_KINDS = Object.keys(RATE_FIELDS) as RateKind[];

// DeepSeek and friends spell the cache-read rate differently.
const CACHE_READ_ALIASES = ["input_cost_per_token_cache_hit"] as const;

// When a model omits a rate, bill those tokens at this rate instead.
const RATE_FALLBACK: Partial<Record<RateKind, RateKind>> = {
  cache_read: "input",
  cache_read_audio: "cache_read",
  cache_creation: "input",
  cache_creation_1h: "cache_creation",
  reasoning: "output",
  input_audio: "input",
  output_audio: "output",
  input_image: "input",
  output_image: "output"
};

// Rate families with above_{N}_tokens variants upstream; reasoning has none.
const THRESHOLD_BASES: Record<string, RateKind> = {
  input_cost_per_token: "input",
  output_cost_per_token: "output",
  cache_read_input_token_cost: "cache_read",
  cache_creation_input_token_cost: "cache_creation",
  cache_creation_input_token_cost_above_1hr: "cache_creation_1h"
};

// Matched against the whole key so unrelated suffixes cannot become tiers.
const THRESHOLD_KEY = new RegExp(
  `^(${Object.keys(THRESHOLD_BASES)
    .sort((a, b) => b.length - a.length)
    .join("|")})_above_(\\d+)(k?)_tokens(?:_(batches|priority|flex))?$`
);

/** Per-token rates for one pricing context. `null` means not declared. */
export interface TokenRates {
  input: Decimal | null;
  output: Decimal | null;
  cache_read: Decimal | null;
  cache_read_audio: Decimal | null;
  cache_creation: Decimal | null;
  cache_creation_audio: Decimal | null;
  cache_creation_1h: Decimal | null;
  reasoning: Decimal | null;
  input_audio: Decimal | null;
  output_audio: Decimal | null;
  input_image: Decimal | null;
  output_image: Decimal | null;
}

export const EMPTY_RATES: TokenRates = {
  input: null,
  output: null,
  cache_read: null,
  cache_read_audio: null,
  cache_creation: null,
  cache_creation_audio: null,
  cache_creation_1h: null,
  reasoning: null,
  input_audio: null,
  output_audio: null,
  input_image: null,
  output_image: null
};

/** Return the rate for `kind`, falling back per `RATE_FALLBACK`. */
export function rateFor(rates: TokenRates, kind: RateKind): Decimal | null {
  const direct = rates[kind];
  if (direct !== null) {
    return direct;
  }
  const fallback = RATE_FALLBACK[kind];
  return fallback === undefined ? null : rateFor(rates, fallback);
}

export interface PricingModeProfile {
  rates: TokenRates;
  thresholds: PricingThreshold[];
}

export interface PricingThreshold {
  threshold: number;
  key: string;
  rates: TokenRates;
}

export interface PricingTier {
  rangeStart: Decimal;
  rangeEnd: Decimal;
  rates: TokenRates;
}

export interface QueryPricingTier {
  rangeStart: number;
  rangeEnd: number;
  rate: Decimal;
}

function toDecimal(value: unknown): Decimal | null {
  if (value === null || value === undefined || typeof value === "boolean") {
    return null;
  }
  if (typeof value !== "number" && typeof value !== "string") {
    return null;
  }
  if (typeof value === "number" && !Number.isFinite(value)) {
    return null;
  }

  let parsed: Decimal;
  try {
    parsed = new Decimal(String(value).trim());
  } catch {
    return null;
  }

  if (!parsed.isFinite() || parsed.lessThan(0)) {
    return null;
  }
  return parsed;
}

/** Read the flat per-token rates a model declares. */
export function parseBaseRates(raw: Record<string, unknown>, suffix = ""): TokenRates {
  const rates: TokenRates = { ...EMPTY_RATES };
  const fieldSuffix = suffix.length > 0 ? `_${suffix}` : "";
  for (const kind of RATE_KINDS) {
    rates[kind] = toDecimal(raw[`${RATE_FIELDS[kind]}${fieldSuffix}`]);
  }

  if (suffix.length === 0 && rates.cache_read === null) {
    for (const alias of CACHE_READ_ALIASES) {
      const aliased = toDecimal(raw[alias]);
      if (aliased !== null) {
        rates.cache_read = aliased;
        break;
      }
    }
  }

  return rates;
}

/** Collect above-N-tokens rates, highest threshold first. */
export function parseThresholds(
  raw: Record<string, unknown>,
  suffix = ""
): PricingThreshold[] {
  const byThreshold = new Map<number, { key: string; rates: TokenRates }>();

  for (const [key, value] of Object.entries(raw)) {
    const match = THRESHOLD_KEY.exec(key);
    if (match === null || (match[4] ?? "") !== suffix) {
      continue;
    }
    const rate = toDecimal(value);
    if (rate === null) {
      continue;
    }

    const amount = Number(match[2]);
    const threshold = match[3] === "k" ? amount * 1000 : amount;
    const entry =
      byThreshold.get(threshold) ??
      { key: `above_${match[2]}${match[3]}_tokens`, rates: { ...EMPTY_RATES } };
    const base = match[1];
    if (base !== undefined) {
      const kind = THRESHOLD_BASES[base];
      if (kind !== undefined) {
        entry.rates[kind] = rate;
      }
    }
    byThreshold.set(threshold, entry);
  }

  return [...byThreshold.entries()]
    .sort((a, b) => b[0] - a[0])
    .map(([threshold, entry]) => ({ threshold, key: entry.key, rates: entry.rates }));
}

export function parsePricingModes(
  raw: Record<string, unknown>
): Partial<Record<PricingMode, PricingModeProfile>> {
  const profiles: Partial<Record<PricingMode, PricingModeProfile>> = {};
  for (const mode of ["batch", "priority", "flex"] as const) {
    const suffix = MODE_SUFFIXES[mode];
    const rates = parseBaseRates(raw, suffix);
    const thresholds = parseThresholds(raw, suffix);
    if (RATE_KINDS.some((kind) => rates[kind] !== null) || thresholds.length > 0) {
      profiles[mode] = { rates, thresholds };
    }
  }
  return profiles;
}

export function overlayRates(base: TokenRates, overrides: TokenRates): TokenRates {
  const resolved = { ...base };
  for (const kind of RATE_KINDS) {
    if (overrides[kind] !== null) {
      resolved[kind] = overrides[kind];
    }
  }
  return resolved;
}

export function resolveModeRates(
  base: TokenRates,
  profile: PricingModeProfile,
  mode: PricingMode,
  standardBase: TokenRates = base,
  provider: string | null = null
): TokenRates {
  if (mode !== "batch") {
    const resolved = overlayRates(base, profile.rates);
    if (profile.rates.output !== null && profile.rates.reasoning === null) {
      resolved.reasoning = null;
    }
    return resolved;
  }
  const referenceInput = standardBase.input ?? base.input;
  const referenceOutput = standardBase.output ?? base.output;
  const inputRatio =
    profile.rates.input !== null && referenceInput !== null && !referenceInput.isZero()
      ? profile.rates.input.div(referenceInput)
      : new Decimal("0.5");
  const outputRatio =
    profile.rates.output !== null && referenceOutput !== null && !referenceOutput.isZero()
      ? profile.rates.output.div(referenceOutput)
      : inputRatio;
  const providerName = provider?.toLowerCase() ?? "";
  const preservesCache = providerName === "gemini" || providerName.startsWith("vertex_ai");
  const cacheRatio = preservesCache ? new Decimal(1) : inputRatio;
  const discounted: TokenRates = {
    ...base,
    input: base.input?.mul(inputRatio) ?? null,
    output: base.output?.mul(outputRatio) ?? null,
    cache_read: base.cache_read?.mul(cacheRatio) ?? null,
    cache_read_audio: base.cache_read_audio?.mul(cacheRatio) ?? null,
    cache_creation: base.cache_creation?.mul(cacheRatio) ?? null,
    cache_creation_audio: base.cache_creation_audio?.mul(cacheRatio) ?? null,
    cache_creation_1h: base.cache_creation_1h?.mul(cacheRatio) ?? null,
    reasoning: null,
    input_audio: base.input_audio?.mul(inputRatio) ?? null,
    output_audio: base.output_audio?.mul(outputRatio) ?? null,
    input_image: base.input_image?.mul(inputRatio) ?? null,
    output_image: base.output_image?.mul(outputRatio) ?? null
  };
  return overlayRates(discounted, { ...profile.rates, input: null, output: null });
}

/**
 * Parse request-size tiers, sorted by range start. Returns [] when unusable.
 *
 * Search-style entries (`input_cost_per_query` with `max_results_range`) carry
 * no `range` and no per-token rate, so they yield [] and leave the model
 * unpriceable rather than being mistaken for token tiers.
 */
export function parseTiers(raw: unknown): PricingTier[] {
  if (!Array.isArray(raw)) {
    return [];
  }
  if (!raw.some((entry) => typeof entry === "object" && entry !== null && "range" in entry)) {
    return [];
  }

  const tiers: PricingTier[] = [];
  for (const entry of raw) {
    if (typeof entry !== "object" || entry === null || !("range" in entry)) {
      return [];
    }
    const record = entry as Record<string, unknown>;
    const bounds = record["range"];
    if (!Array.isArray(bounds) || bounds.length !== 2) {
      return [];
    }
    const start = toDecimal(bounds[0]);
    const end = toDecimal(bounds[1]);
    if (start === null || end === null || end.lessThanOrEqualTo(start)) {
      return [];
    }
    const rates = parseBaseRates(record);
    if (rates.input === null) {
      return [];
    }
    tiers.push({ rangeStart: start, rangeEnd: end, rates });
  }

  tiers.sort((a, b) => a.rangeStart.comparedTo(b.rangeStart));
  for (let index = 1; index < tiers.length; index += 1) {
    const previous = tiers[index - 1];
    const current = tiers[index];
    if (
      previous !== undefined &&
      current !== undefined &&
      current.rangeStart.lessThan(previous.rangeEnd)
    ) {
      return [];
    }
  }
  return tiers;
}

export function parseQueryTiers(raw: unknown): QueryPricingTier[] {
  if (!Array.isArray(raw)) {
    return [];
  }
  if (!raw.some((entry) =>
    typeof entry === "object" && entry !== null && "max_results_range" in entry
  )) {
    return [];
  }

  const tiers: QueryPricingTier[] = [];
  for (const entry of raw) {
    if (typeof entry !== "object" || entry === null) {
      return [];
    }
    const record = entry as Record<string, unknown>;
    const bounds = record["max_results_range"];
    const rate = toDecimal(record["input_cost_per_query"]);
    if (
      !Array.isArray(bounds) ||
      bounds.length !== 2 ||
      !Number.isSafeInteger(bounds[0]) ||
      !Number.isSafeInteger(bounds[1]) ||
      (bounds[0] as number) < 0 ||
      (bounds[1] as number) < (bounds[0] as number) ||
      rate === null
    ) {
      return [];
    }
    tiers.push({ rangeStart: bounds[0] as number, rangeEnd: bounds[1] as number, rate });
  }

  tiers.sort((a, b) => a.rangeStart - b.rangeStart);
  for (let index = 1; index < tiers.length; index += 1) {
    const previous = tiers[index - 1];
    const current = tiers[index];
    if (previous !== undefined && current !== undefined && current.rangeStart <= previous.rangeEnd) {
      return [];
    }
  }
  return tiers;
}

export function selectQueryTierRate(
  tiers: readonly QueryPricingTier[],
  maxResults: number
): Decimal | null {
  return tiers.find(
    (tier) => tier.rangeStart <= maxResults && maxResults <= tier.rangeEnd
  )?.rate ?? null;
}

/**
 * Apply the highest matching threshold on top of the base rates.
 *
 * The trigger is the input token count alone. xAI includes the exact boundary;
 * other providers require the count to exceed it. A threshold that declares
 * only some rates leaves the rest on their base values.
 */
export function resolveRates(
  base: TokenRates,
  thresholds: readonly PricingThreshold[],
  inputTokens: number,
  provider: string | null = null
): [TokenRates, string | null] {
  const inclusive = provider?.toLowerCase() === "xai";
  for (const threshold of thresholds) {
    if (inputTokens > threshold.threshold || (inclusive && inputTokens === threshold.threshold)) {
      const resolved: TokenRates = { ...base };
      for (const kind of RATE_KINDS) {
        const override = threshold.rates[kind];
        if (override !== null) {
          resolved[kind] = override;
        }
      }
      if (threshold.rates.output !== null && threshold.rates.reasoning === null) {
        resolved.reasoning = null;
      }
      return [resolved, threshold.key];
    }
  }
  return [base, null];
}

/** Select the request-wide rates for a total input token count. */
export function selectTierRates(
  tiers: readonly PricingTier[],
  inputTokens: number
): TokenRates | null {
  if (tiers.length === 0) {
    return null;
  }

  const tokenCount = new Decimal(inputTokens);
  if (tokenCount.lessThanOrEqualTo(0)) {
    return null;
  }
  for (const tier of tiers) {
    if (
      tokenCount.greaterThan(tier.rangeStart) &&
      tokenCount.lessThanOrEqualTo(tier.rangeEnd)
    ) {
      return tier.rates;
    }
  }

  const last = tiers[tiers.length - 1];
  return last?.rates ?? null;
}

export function resolveTierRates(base: TokenRates, tier: TokenRates): TokenRates {
  const tierDeclaresOutput = tier.output !== null;
  return {
    input: tier.input,
    output: tier.output ?? base.output,
    cache_read: tier.cache_read,
    cache_read_audio: tier.cache_read_audio,
    cache_creation: tier.cache_creation,
    cache_creation_audio: tier.cache_creation_audio,
    cache_creation_1h: tier.cache_creation_1h,
    reasoning: tier.reasoning ?? (tierDeclaresOutput ? null : base.reasoning),
    input_audio: tier.input_audio,
    output_audio: tier.output_audio,
    input_image: tier.input_image,
    output_image: tier.output_image
  };
}

File: llmcalc/pricing_tiers.py
"""Tiered, long-context, cache and reasoning pricing resolution.

Pure helpers: no I/O, no cache, no network. Upstream ships two tiering
mechanisms that behave differently, so both are modelled separately:

* `above_{N}_tokens` rates replace the base rates for the whole request once the
  input token count exceeds the threshold, and apply to output as well as input.
* `tiered_pricing` selects one rate table from the request's total input size.

Token, cache, reasoning, audio and image rates are tracked independently.
Missing specialized rates fall back through the corresponding input or output
rate, except one-hour cache writes, which callers require explicitly.

Parsers return empty results for unsupported tier shapes. The pricing parser
reports malformed supported fields through structured diagnostics.
"""

from __future__ import annotations

import re
from collections.abc import Mapping, Sequence
from dataclasses import dataclass, replace
from decimal import Decimal, InvalidOperation
from typing import Any, Literal

RateKind = Literal[
    "input",
    "output",
    "cache_read",
    "cache_read_audio",
    "cache_creation",
    "cache_creation_audio",
    "cache_creation_1h",
    "reasoning",
    "input_audio",
    "output_audio",
    "input_image",
    "output_image",
]
PricingMode = Literal["standard", "batch", "priority", "flex"]

# Upstream field name for each rate kind.
RATE_FIELDS: dict[RateKind, str] = {
    "input": "input_cost_per_token",
    "output": "output_cost_per_token",
    "cache_read": "cache_read_input_token_cost",
    "cache_read_audio": "cache_read_input_audio_token_cost",
    "cache_creation": "cache_creation_input_token_cost",
    "cache_creation_audio": "cache_creation_input_audio_token_cost",
    "cache_creation_1h": "cache_creation_input_token_cost_above_1hr",
    "reasoning": "output_cost_per_reasoning_token",
    "input_audio": "input_cost_per_audio_token",
    "output_audio": "output_cost_per_audio_token",
    "input_image": "input_cost_per_image_token",
    "output_image": "output_cost_per_image_token",
}

MODE_SUFFIXES: dict[PricingMode, str] = {
    "standard": "",
    "batch": "batches",
    "priority": "priority",
    "flex": "flex",
}

# DeepSeek and friends spell the cache-read rate differently.
_CACHE_READ_ALIASES = ("input_cost_per_token_cache_hit",)

# When a model omits a rate, bill those tokens at this rate instead.
_RATE_FALLBACK: dict[RateKind, RateKind] = {
    "cache_read": "input",
    "cache_read_audio": "cache_read",
    "cache_creation": "input",
    "cache_creation_1h": "cache_creation",
    "reasoning": "output",
    "input_audio": "input",
    "output_audio": "output",
    "input_image": "input",
    "output_image": "output",
}

# Rate families with above_{N}_tokens variants upstream; reasoning has none.
_THRESHOLD_BASES: dict[str, RateKind] = {
    "input_cost_per_token": "input",
    "output_cost_per_token": "output",
    "cache_read_input_token_cost": "cache_read",
    "cache_creation_input_token_cost": "cache_creation",
    "cache_creation_input_token_cost_above_1hr": "cache_creation_1h",
}

_THRESHOLD_KEY = re.compile(
    r"^(?P<base>"
    + "|".join(sorted(_THRESHOLD_BASES, key=len, reverse=True))
    + r")_above_(?P<amount>\d+)(?P<kilo>k?)_tokens"
    + r"(?:_(?P<suffix>batches|priority|flex))?$"
)

_ZERO = Decimal("0")


@dataclass(frozen=True)
class TokenRates:
    """Per-token rates for one pricing context. `None` means not declared."""

    input: Decimal | None = None
    output: Decimal | None = None
    cache_read: Decimal | None = None
    cache_read_audio: Decimal | None = None
    cache_creation: Decimal | None = None
    cache_creation_audio: Decimal | None = None
    cache_creation_1h: Decimal | None = None
    reasoning: Decimal | None = None
    input_audio: Decimal | None = None
    output_audio: Decimal | None = None
    input_image: Decimal | None = None
    output_image: Decimal | None = None

    def get(self, kind: RateKind) -> Decimal | None:
        """Return the rate for `kind`, falling back per `_RATE_FALLBACK`."""
        direct: Decimal | None = getattr(self, kind)
        if direct is not None:
            return direct
        fallback = _RATE_FALLBACK.get(kind)
        return None if fallback is None else self.get(fallback)


@dataclass(frozen=True)
class PricingModeProfile:
    rates: TokenRates
    thresholds: tuple[PricingThreshold, ...]

@dataclass(frozen=True)
class PricingThreshold:
    threshold: int
    key: str
    rates: TokenRates


@dataclass(frozen=True)
class PricingTier:
    range_start: Decimal
    range_end: Decimal
    rates: TokenRates


@dataclass(frozen=True)
class QueryPricingTier:
    range_start: int
    range_end: int
    rate: Decimal


def _to_decimal(value: Any) -> Decimal | None:
    if value is None or isinstance(value, bool):
        return None
    if isinstance(value, Decimal):
        return value if value >= _ZERO else None
    if isinstance(value, (int, float, str)):
        try:
            parsed = Decimal(str(value).strip())
        except (InvalidOperation, ValueError):
            return None
        if not parsed.is_finite() or parsed < _ZERO:
            return None
        return parsed
    return None


def parse_base_rates(raw: Mapping[str, Any], suffix: str = "") -> TokenRates:
    """Read the flat per-token rates a model declares."""
    field_suffix = f"_{suffix}" if suffix else ""
    values: dict[str, Decimal | None] = {
        kind: _to_decimal(raw.get(f"{field}{field_suffix}"))
        for kind, field in RATE_FIELDS.items()
    }

    if not suffix and values["cache_read"] is None:
        for alias in _CACHE_READ_ALIASES:
            aliased = _to_decimal(raw.get(alias))
            if aliased is not None:
                values["cache_read"] = aliased
                break

    return TokenRates(**values)


def parse_thresholds(
    raw: Mapping[str, Any], suffix: str = ""
) -> tuple[PricingThreshold, ...]:
    """Collect above-N-tokens rates, highest threshold first."""
    by_threshold: dict[int, dict[str, Any]] = {}

    for key, value in raw.items():
        match = _THRESHOLD_KEY.match(key)
        if match is None or (match.group("suffix") or "") != suffix:
            continue
        rate = _to_decimal(value)
        if rate is None:
            continue
        amount = int(match.group("amount"))
        threshold = amount * 1000 if match.group("kilo") else amount
        entry = by_threshold.setdefault(
            threshold,
            {"key": f"above_{match.group('amount')}{match.group('kilo')}_tokens", "rates": {}},
        )
        entry["rates"][_THRESHOLD_BASES[match.group("base")]] = rate

    return tuple(
        PricingThreshold(
            threshold=threshold,
            key=entry["key"],
            rates=TokenRates(**entry["rates"]),
        )
        for threshold, entry in sorted(by_threshold.items(), reverse=True)
    )


def parse_pricing_modes(raw: Mapping[str, Any]) -> dict[PricingMode, PricingModeProfile]:
    profiles: dict[PricingMode, PricingModeProfile] = {}
    for mode, suffix in MODE_SUFFIXES.items():
        if mode == "standard":
            continue
        rates = parse_base_rates(raw, suffix)
        thresholds = parse_thresholds(raw, suffix)
        if any(getattr(rates, kind) is not None for kind in RATE_FIELDS) or thresholds:
            profiles[mode] = PricingModeProfile(rates=rates, thresholds=thresholds)
    return profiles


def overlay_rates(base: TokenRates, overrides: TokenRates) -> TokenRates:
    values: dict[str, Decimal | None] = {
        kind: getattr(overrides, kind)
        if getattr(overrides, kind) is not None
        else getattr(base, kind)
        for kind in RATE_FIELDS
    }
    return TokenRates(**values)


def resolve_mode_rates(
    base: TokenRates,
    profile: PricingModeProfile,
    mode: PricingMode,
    *,
    standard_base: TokenRates | None = None,
    provider: str | None = None,
) -> TokenRates:
    if mode != "batch":
        resolved = overlay_rates(base, profile.rates)
        if profile.rates.output is not None and profile.rates.reasoning is None:
            resolved = replace(resolved, reasoning=None)
        return resolved

    reference = standard_base or base
    reference_input = reference.input if reference.input is not None else base.input
    reference_output = reference.output if reference.output is not None else base.output
    input_ratio = (
        profile.rates.input / reference_input
        if profile.rates.input is not None
        and reference_input is not None
        and reference_input != _ZERO
        else Decimal("0.5")
    )
    output_ratio = (
        profile.rates.output / reference_output
        if profile.rates.output is not None
        and reference_output is not None
        and reference_output != _ZERO
        else input_ratio
    )
    provider_name = (provider or "").casefold()
    preserves_cache = provider_name == "gemini" or provider_name.startswith("vertex_ai")
    cache_ratio = Decimal("1") if preserves_cache else input_ratio
    discounted = replace(
        base,
        input=None if base.input is None else base.input * input_ratio,
        output=None if base.output is None else base.output * output_ratio,
        cache_read=(
            None if base.cache_read is None else base.cache_read * cache_ratio
        ),
        cache_read_audio=(
            None
            if base.cache_read_audio is None
            else base.cache_read_audio * cache_ratio
        ),
        cache_creation=(
            None
            if base.cache_creation is None
            else base.cache_creation * cache_ratio
        ),
        cache_creation_audio=(
            None
            if base.cache_creation_audio is None
            else base.cache_creation_audio * cache_ratio
        ),
        cache_creation_1h=(
            None
            if base.cache_creation_1h is None
            else base.cache_creation_1h * cache_ratio
        ),
        reasoning=None,
        input_audio=None if base.input_audio is None else base.input_audio * input_ratio,
        output_audio=None if base.output_audio is None else base.output_audio * output_ratio,
        input_image=None if base.input_image is None else base.input_image * input_ratio,
        output_image=None if base.output_image is None else base.output_image * output_ratio,
    )
    batch_overrides = replace(profile.rates, input=None, output=None)
    return overlay_rates(discounted, batch_overrides)


def parse_tiers(raw: Any) -> tuple[PricingTier, ...]:
    """Parse request-size tiers, sorted by range start. Returns () when unusable.

    Search-style entries (`input_cost_per_query` with `max_results_range`) carry
    no `range` and no per-token rate, so they yield () and leave the model
    unpriceable rather than being mistaken for token tiers.
    """
    if not isinstance(raw, (list, tuple)):
        return ()

    if not any(isinstance(entry, Mapping) and "range" in entry for entry in raw):
        return ()

    tiers: list[PricingTier] = []
    for entry in raw:
        if not isinstance(entry, Mapping) or "range" not in entry:
            return ()
        bounds = entry.get("range")
        if not isinstance(bounds, (list, tuple)) or len(bounds) != 2:
            return ()
        start = _to_decimal(bounds[0])
        end = _to_decimal(bounds[1])
        if start is None or end is None or end <= start:
            return ()
        rates = parse_base_rates(entry)
        if rates.input is None:
            return ()
        tiers.append(PricingTier(range_start=start, range_end=end, rates=rates))

    tiers.sort(key=lambda tier: tier.range_start)
    if any(
        current.range_start < previous.range_end
        for previous, current in zip(tiers, tiers[1:], strict=False)
    ):
        return ()
    return tuple(tiers)


def parse_query_tiers(raw: Any) -> tuple[QueryPricingTier, ...]:
    if not isinstance(raw, (list, tuple)):
        return ()
    if not any(
        isinstance(entry, Mapping) and "max_results_range" in entry for entry in raw
    ):
        return ()

    tiers: list[QueryPricingTier] = []
    for entry in raw:
        if not isinstance(entry, Mapping):
            return ()
        bounds = entry.get("max_results_range")
        rate = _to_decimal(entry.get("input_cost_per_query"))
        if (
            not isinstance(bounds, (list, tuple))
            or len(bounds) != 2
            or type(bounds[0]) is not int
            or type(bounds[1]) is not int
            or bounds[0] < 0
            or bounds[1] < bounds[0]
            or rate is None
        ):
            return ()
        tiers.append(QueryPricingTier(bounds[0], bounds[1], rate))

    tiers.sort(key=lambda tier: tier.range_start)
    if any(
        current.range_start <= previous.range_end
        for previous, current in zip(tiers, tiers[1:], strict=False)
    ):
        return ()
    return tuple(tiers)


def select_query_tier_rate(
    tiers: Sequence[QueryPricingTier], max_results: int
) -> Decimal | None:
    for tier in tiers:
        if tier.range_start <= max_results <= tier.range_end:
            return tier.rate
    return None


def resolve_rates(
    base: TokenRates,
    thresholds: Sequence[PricingThreshold],
    input_tokens: int,
    provider: str | None = None,
) -> tuple[TokenRates, str | None]:
    """Apply the highest matching threshold on top of the base rates.

    The trigger is the input token count alone. xAI includes the exact boundary;
    other providers require the count to exceed it. A threshold that declares
    only some rates leaves the rest on their base values.
    """
    inclusive = provider is not None and provider.casefold() == "xai"
    for threshold in thresholds:
        if input_tokens > threshold.threshold or (
            inclusive and input_tokens == threshold.threshold
        ):
            overrides: dict[str, Decimal | None] = {}
            for kind in RATE_FIELDS:
                value = getattr(threshold.rates, kind)
                if value is not None:
                    overrides[str(kind)] = value
            if threshold.rates.output is not None and threshold.rates.reasoning is None:
                overrides["reasoning"] = None
            return replace(base, **overrides), threshold.key
    return base, None


def select_tier_rates(
    tiers: Sequence[PricingTier], input_tokens: int
) -> TokenRates | None:
    """Select the request-wide rates for a total input token count."""
    if not tiers:
        return None

    token_count = Decimal(input_tokens)
    for tier in tiers:
        if tier.range_start < token_count <= tier.range_end:
            return tier.rates

    if token_count > 0:
        return tiers[-1].rates
    return None


def resolve_tier_rates(base: TokenRates, tier: TokenRates) -> TokenRates:
    """Resolve a selected tier with LiteLLM's per-kind fallback semantics."""
    tier_declares_output = tier.output is not None
    return TokenRates(
        input=tier.input,
        output=tier.output if tier.output is not None else base.output,
        cache_read=tier.cache_read,
        cache_read_audio=tier.cache_read_audio,
        cache_creation=tier.cache_creation,
        cache_creation_audio=tier.cache_creation_audio,
        cache_creation_1h=tier.cache_creation_1h,
        reasoning=(
            tier.reasoning
            if tier.reasoning is not None
            else None
            if tier_declares_output
            else base.reasoning
        ),
        input_audio=tier.input_audio,
        output_audio=tier.output_audio,
        input_image=tier.input_image,
        output_image=tier.output_image,
    )
